### ISSUE TITLE
C# Implementation

### ISSUE THREAD

**Issue by Draco18s** [wbrown/novelai-research-tool#1]
Saw your project and converted it to C# (my preferred language). Just thought I'd drop by and leave this reference link for future explorers, as it's not forked per git logic.

https://github.com/Draco18s/NovelAPI

I'm a bit behind and some of your latest commits are mighty fine. 👍

**Comment by wbrown** [comment]
Nice! You probably noticed that I just added a context builder with lorebook, A/N, and memory support, as well as  `.scenario` file support, though I haven't exposed it.

Project is a bit of a moving target at the moment. :)

**Comment by Draco18s** [comment]
Yes I did! And no worries about the moving target. I'm probably going to be handling authors note, lore, etc. via a separate mechanism (although latching onto NovelAI's storage and/or format would be handy), my project (at least, for today) was just to get a response back from the AI.

I've basically been building a discord bot for myself and some friends, found this project and figured what the heck. NovelAI doesn't have multiplayer yet, so this might be a close second functionality wise.

Thank goodness Go's syntax is close enough such that, while I have to translate sections by hand, I can at least *read* the code and not have to do too much guessing. And the project isn't that large. There's apparently a [converter(https://github.com/GridProtectionAlliance/go2cs), but I couldn't get it to work.

**Comment by wbrown** [comment]
Yeah, different intention.  `nrt` is supposed to be a close simulation of how NovelAI interacts with the user, so I'm trying to get reasonably close to the context builder model.  You might try the `text_adventure` prefix if you want to be interacting with it like an adventure.  Just give it `> ` prompts, like `adventure.go` does.  I would also allow a couple-few iterations of generations before returning a response so that you can get a meaningful response.

**Comment by Draco18s** [comment]
Oh for sure! And I'll be adding that functionality too. As what I'd like to have is an API that is feature-rich enough that it's useful, while also suiting my own needs. Just that *how* lore and such is stored and referenced is going to be a challenge, given the nature of the front end I'm working with (I'm quite pleased with using reactions as undo/reroll/continue buttons though!)

And yeah, longer generation/more iterations before returning it to the player is a good idea.

**Comment by Draco18s** [comment]
Still having trouble with the auth part though. I know what the hashes *should* be (after extracting them from Chrome's dev panel), and I've found the JS that generates them, but I can't get my code to produce them. I've tried all sorts of things regarding the various parameters, including finding the Sodium library and translating ITS code to C#. You've apparently gotten yours to work, so any insights?

https://github.com/Draco18s/NovelAPI/blob/main/Auth.cs#L36-L50

**Comment by Draco18s** [comment]
Just thought I'd highlight that I got the Keystore endpoint worked out.
https://github.com/Draco18s/NovelAPI/blob/main/authentication/Auth.cs#L137

With this the various other encrypted endpoints (server-stored stories, ai modules, etc) can be fetched and decrypted as each story/module has its own secret key.

I believe that a call to user/objects/stories will return a list of stories by meta-ID, and that meta-ID is how to look up the encryption secret key in the keystore, though I haven't yet worked out how to turn that meta-ID into a url to fetch the story content from.

Similarly the keystore has keys for accessing the AI modules the user has uploaded (fetched from user/objects/aimodules and comes back as a list of encrypted json objects).
